File: app/src/main/java/com/left/gank/ui/gallery/GalleryFragment.java
package com.left.gank.ui.gallery;

import android.os.Bundle;
import android.ui.logcat.Logcat;
import android.view.View;

import com.google.android.material.bottomsheet.BottomSheetDialog;
import com.left.gank.R;
import com.left.gank.rx.RxSchedulers;
import com.left.gank.ui.base.fragment.SupportFragment;
import com.left.gank.utils.RxSaveImage;
import com.left.gank.utils.ShareUtils;
import com.left.gank.widget.BottomSheetGalleryMenuDialog;
import com.left.gank.widget.ProgressImageView;

import java.io.File;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import butterknife.BindView;

/**
 * 图片浏览
 * Create by LingYan on 2016-12-19
 */
public class GalleryFragment extends SupportFragment {
    private static final String IMAGE_URL = "Image_Url";

    @BindView(R.id.progress_image)
    ProgressImageView progressImage;

    private String url;

    @Override
    protected int fragmentLayoutId() {
        return R.layout.fragment_browse_picture;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Bundle bundle = getArguments();
        if (bundle != null) {
            url = bundle.getString(IMAGE_URL);
        }
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        progressImage.load(url, getContext());
        progressImage.setImageViewOnClick(imageViewOnClick);
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
    }

    private final ProgressImageView.ImageViewOnClick imageViewOnClick = new ProgressImageView.ImageViewOnClick() {
        @Override
        public void onImageClick(View v) {
            BottomSheetGalleryMenuDialog bottomSheetGalleryMenuDialog = new BottomSheetGalleryMenuDialog(getActivity());
            bottomSheetGalleryMenuDialog.setCallback(callback);
            bottomSheetGalleryMenuDialog.show();
        }

        @Override
        public void onLongClick(View v) {
            BottomSheetDialog bottomSheetDialog = new BottomSheetDialog(getActivity());
            bottomSheetDialog.setContentView(R.layout.view_gallery_menu);
            bottomSheetDialog.show();
        }
    };

    private final BottomSheetGalleryMenuDialog.Callback callback = new BottomSheetGalleryMenuDialog.Callback() {
        @Override
        public void share() {
            RxSaveImage.convertBitmap(getContext(), url)
                    .observeOn(RxSchedulers.INSTANCE.mainThread())
                    .map(bitmap -> {
                        File file = RxSaveImage.createImageFile(String.valueOf(bitmap.hashCode()));
                        if (file != null) {
                            return RxSaveImage.bitmapToImage(getContext(), bitmap, file);
                        }
                        return null;
                    })
                    .as(bindLifecycle())
                    .subscribe(uri -> ShareUtils.shareSingleImage(getContext(), uri), Logcat::e);
        }

        @Override
        public void save() {
            RxSaveImage.convertBitmap(getContext(), url)
                    .observeOn(RxSchedulers.INSTANCE.mainThread())
                    .map(bitmap -> {
                        File file = RxSaveImage.createImageFile(String.valueOf(bitmap.hashCode()));
                        if (file != null) {
                            return RxSaveImage.bitmapToImage(getContext(), bitmap, file);
                        }
                        return null;
                    })
                    .as(bindLifecycle())
                    .subscribe(uri -> shortToast(uri.toString()), Logcat::e);
        }
    };

    public static GalleryFragment newInstance(String url) {
        GalleryFragment fragment = new GalleryFragment();
        Bundle args = new Bundle();
        args.putString(IMAGE_URL, url);
        fragment.setArguments(args);
        return fragment;
    }
}

File: app/src/main/java/com/left/gank/ui/gallery/GalleryFragment.kt
package com.left.gank.ui.gallery

import android.business.domain.Gift
import android.net.Uri
import android.os.Bundle
import android.ui.logcat.Logcat
import android.view.View
import com.left.gank.R
import com.left.gank.base.fragment.SupportFragment
import com.left.gank.rx.RxSchedulers
import com.left.gank.utils.RxSaveImage
import com.left.gank.utils.ShareUtils
import com.left.gank.widget.BottomSheetGalleryMenuDialog
import com.left.gank.widget.ProgressImageView
import com.uber.autodispose.ObservableSubscribeProxy
import kotlinx.android.synthetic.main.fragment_browse_picture.*

/**
 * 图片浏览
 * Create by LingYan on 2016-12-19
 */
class GalleryFragment : SupportFragment() {
    private var url: String? = null
    private var postUrl: String? = null

    override fun fragmentLayoutId(): Int {
        return R.layout.fragment_browse_picture
    }

    private val progressCallback = object : ProgressImageView.ImageCallback {
        override fun onImageClick(v: View) {

        }

        override fun onLongClick(v: View) {
            BottomSheetGalleryMenuDialog(activity!!).apply {
                setCallback(callback)
                show()
            }
        }
    }

    private val callback = object : BottomSheetGalleryMenuDialog.Callback {
        override fun share() {
            onEvent(false)
        }

        override fun save() {
            onEvent(true)
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        arguments?.apply {
            url = getString(IMAGE_URL)
            postUrl = getString(IMAGE_POST_URL)
        }
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        progress_image!!.apply {
            load(url, postUrl, context)
            setImageViewOnClick(progressCallback)
        }
    }

    private fun onEvent(save: Boolean) {
        RxSaveImage.convertBitmap(context, url)
                .observeOn(RxSchedulers.mainThread())
                .map<Uri> { bitmap ->
                    val file = RxSaveImage.createImageFile(bitmap.hashCode().toString())
                    if (file != null) {
                        return@map RxSaveImage.bitmapToImage(context, bitmap, file)
                    }
                    null
                }
                .`as`<ObservableSubscribeProxy<Uri>>(bindLifecycle<Uri>())
                .subscribe({ uri -> if (save) shortToast(uri.toString()) else ShareUtils.shareSingleImage(context!!, uri) }, { throwable -> Logcat.e(throwable) })

    }

    companion object {
        const val IMAGE_URL = "Image_Url"
        const val IMAGE_POST_URL = "Post_Url"

        fun newInstance(gift: Gift): GalleryFragment = GalleryFragment().apply {
            arguments = Bundle().apply {
                putString(IMAGE_URL, gift.imgUrl)
                putString(IMAGE_POST_URL, gift.url)
            }
        }
    }
}
